mapper/tedge_mapper/src/mapper.rs

```diff
@@ -86,26 +91,103 @@ impl Mapper {
 
     #[instrument(skip(self), name = "messages")]
     async fn subscribe_messages(&self) -> Result<(), MqttClientError> {
-        let mut messages = self.client.subscribe(self.config.in_topic.filter()).await?;
+        let mut messages = self
+            .client
+            .subscribe(self.config.in_topic_filter.clone())
+            .await?;
+
+        let mut children: HashSet<String> = HashSet::new();
+
         while let Some(message) = messages.next().await {
             debug!("Mapping {:?}", message.payload_str());
-            match self.converter.convert(message.payload_str()?) {
-                Ok(mapped) => {
-                    self.client
-                        .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
-                        .await?;
-                }
-                Err(error) => {
-                    debug!("Mapping error: {}", error);
+            debug!("Topic {:?}", message.topic.name);
+
+            // ? Validation for child device ID?
+            match get_child_id_from_topic(message.clone().topic.name) {
+                Some(child_id) if child_id.is_empty() => {
+                    let error = MapperError::InvalidChildId { id: child_id };
+                    error!("Child ID error: {}", error);
                     self.client
                         .publish(mqtt_client::Message::new(
                             &self.config.errors_topic,
                             error.to_string(),
                         ))
                         .await?;
                 }
+                Some(child_id) => {
```

Comment:
I'm wondering if this main mapper loop can be simplified to just a call to a convertor and publish the resultant message back to the MQTT broker, and this whole child device specific checks and manipulations can be moved as the responsibility of the `convertor`. That way this mapper will act just like a message router and the translation logic completely stays in the `convertor` which might make it easier for us to unit test quite a lot of this logic which might have to be integration tested otherwise.

For that, the convertor's convert logic will have to be changed to consume a `Message` and convert it to another `Message` as follows:
```
fn convert(&self, input: &Message) -> Result<Message, Self::Error>;
```
This might eliminate the need for two separate methods `convert()` and `convert_child_device_payload()` here.

Reply:
The signature of `convert` might even have to return a `Vec<Message>` instead of a single `Message` to handle the case of one inbound message resulting in the creation of multiple outbound messages as in the case of this child-device measurement handling:
```
fn convert(&self, input: &Message) -> Result<Vec<Message>, Self::Error>;
```

Reply:
The signature proposed by @albinsuresh has the huge benefit to be prepared to many use cases.

A mapper might return 0, 1, 2, or more messages for an input message.